streams/src/main/java/org/apache/kafka/streams/kstream/internals/suppress/InMemoryTimeOrderedKeyValueBuffer.java

```diff
@@ -0,0 +1,143 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements. See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License. You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.kafka.streams.kstream.internals.suppress;
+
+import org.apache.kafka.common.utils.Bytes;
+import org.apache.kafka.streams.KeyValue;
+import org.apache.kafka.streams.state.internals.ContextualRecord;
+
+import java.util.HashMap;
+import java.util.Iterator;
+import java.util.Map;
+import java.util.TreeMap;
+import java.util.function.Consumer;
+import java.util.function.Supplier;
+
+class InMemoryTimeOrderedKeyValueBuffer implements TimeOrderedKeyValueBuffer<Bytes, ContextualRecord<byte[]>> {
+    private final Map<Bytes, TimeKey<Bytes>> index = new HashMap<>();
+    private final TreeMap<TimeKey<Bytes>, ContextualRecord<byte[]>> sortedMap = new TreeMap<>();
+    private long memBufferSize = 0L;
+    private long minTimestamp = Long.MAX_VALUE;
+
+    private class InnerIterator implements Iterator<KeyValue<TimeKey<Bytes>, ContextualRecord<byte[]>>> {
+        private final Iterator<Map.Entry<TimeKey<Bytes>, ContextualRecord<byte[]>>> delegate = sortedMap.entrySet().iterator();
+        private Map.Entry<TimeKey<Bytes>, ContextualRecord<byte[]>> next;
+        private Map.Entry<TimeKey<Bytes>, ContextualRecord<byte[]>> last;
+        private int nextCount = 0;
+
+        @Override
+        public boolean hasNext() {
+            return delegate.hasNext();
+        }
+
+        @Override
+        public KeyValue<TimeKey<Bytes>, ContextualRecord<byte[]>> next() {
+            final Map.Entry<TimeKey<Bytes>, ContextualRecord<byte[]>> next =
+                this.next != null ? this.next : delegate.next();
+            this.next = null;
+            last = next;
+            nextCount++;
+            return next.getValue() == null ? null : new KeyValue<>(next.getKey(), next.getValue());
+        }
+
+        @Override
+        public void remove() {
+            if (nextCount != 1) {
```

Comment:
yes. This is an optimization to support maximal efficiency in:
* removing some unknown number records, each of which is currently the minimum in the buffer when it gets removed
* maintaining a correct value of `minTimestamp`.

As far as we know right now, we will only ever need to remove the min records from the buffer. I.e., I don't think we need to iterate for a while and *then* remove.

But we may need to remove more than one record, and we won't know if we need to remove the *next* record until after we remove *this* record.

Previously, I didn't have this guard, but in that case, we can't just set `minTimestamp` to the buffer time of the next record upon removing. Because we don't know whether the record we just removed is the leftmost record in the tree without traversing it again. Because of that, I had to avoid updating `minTimestamp` until you close the iterator (and therefore it had to be a `CloseableIterator`). This means that the KTableSuppressProcessor couldn't just keep popping records while the minTimestamp was less than the desired boundary, it had to get the "buffer time" from the TimeKey and make its decision from that.

All in all, it's way cleaner this way, with the expense of that one extra guard.

I could go one step further and make it like a "predicated, consuming iterator", which just pops records out as long as the predicate condition is true. Do you think this would be more straightforward?

Reply:
From your later comments, it seems like you would say it would be more straightforward. I'll go ahead and simplify it.